File: src/python/bsk_stack.py
class BSKStack:
    """
    Бело-Сине-Красный стек:
    White zone  – новые элементы
    Blue zone   – зрелые элементы
    Red zone    – старые, защищённые элементы
    """

    def __init__(self, white_ratio=0.3, blue_ratio=0.4):
        self.data = []
        self.white_ratio = white_ratio
        self.blue_ratio = blue_ratio

    # ----------------------------
    # Внутренние зоны
    # ----------------------------
    def _zones(self):
        n = len(self.data)
        w = int(n * self.white_ratio)
        b = int(n * (self.white_ratio + self.blue_ratio))
        return w, b  # границы: [0:w]=W, [w:b]=B, [b:n]=R

    # ----------------------------
    # Публичные операции
    # ----------------------------
    def push(self, item):
        self.data.insert(0, ("W", item))

    def pop(self):
        if not self.data:
            raise IndexError("BSKStack is empty")

        w, b = self._zones()

        # белая зона
        for i in range(0, w):
            if self.data[i][0] in ("W", "B"):
                return self.data.pop(i)[1]

        # синяя зона
        for i in range(w, b):
            if self.data[i][0] == "B":
                return self.data.pop(i)[1]

        # красная зона — запрещена
        raise PermissionError("Red zone is locked — cannot pop older elements")

    def update_colors(self):
        w, b = self._zones()
        for i, (_, item) in enumerate(self.data):
            if i < w:
                self.data[i] = ("W", item)
            elif i < b:
                self.data[i] = ("B", item)
            else:
                self.data[i] = ("R", item)

    def __len__(self):
        return len(self.data)

    def __repr__(self):
        return str(self.data)

File: src/java/BSKStack.java
import java.util.ArrayList;
import java.util.List;

public class BSKStack<T> {

    private static class Node<T> {
        String color;
        T value;
        Node(String c, T v) { color = c; value = v; }
    }

    private final List<Node<T>> data = new ArrayList<>();
    private final double whiteRatio;
    private final double blueRatio;

    public BSKStack(double whiteRatio, double blueRatio) {
        this.whiteRatio = whiteRatio;
        this.blueRatio = blueRatio;
    }

    public BSKStack() {
        this(0.3, 0.4);
    }

    private int[] zones() {
        int n = data.size();
        int w = (int)(n * whiteRatio);
        int b = (int)(n * (whiteRatio + blueRatio));
        return new int[] { w, b };
    }

    public void push(T item) {
        data.add(0, new Node<>("W", item));
    }

    public T pop() {
        if (data.isEmpty())
            throw new IllegalStateException("Empty BSKStack");

        int[] z = zones();
        int w = z[0], b = z[1];

        // White zone
        for (int i = 0; i < w; i++) {
            if (data.get(i).color.equals("W") || data.get(i).color.equals("B")) {
                return data.remove(i).value;
            }
        }

        // Blue zone
        for (int i = w; i < b; i++) {
            if (data.get(i).color.equals("B")) {
                return data.remove(i).value;
            }
        }

        throw new RuntimeException("Red zone is locked");
    }

    public void updateColors() {
        int[] z = zones();
        int w = z[0], b = z[1];

        for (int i = 0; i < data.size(); i++) {
            Node<T> n = data.get(i);
            if (i < w) n.color = "W";
            else if (i < b) n.color = "B";
            else n.color = "R";
        }
    }

    public int size() { return data.size(); }

    @Override
    public String toString() {
        return data.toString();
    }
}
